fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java

```diff
@@ -73,10 +84,54 @@ protected IModLocator.ModFileOrException createMod(Path path, boolean ignoreUnkn
         } else
             return new IModLocator.ModFileOrException(null, new ModFileLoadingException("Invalid mod file found " + path));
 
-        mjm.setModFile(mod);
         return new IModLocator.ModFileOrException(mod, null);
     }
 
+    private static JarMetadata loadMetaFromJar(SecureJar jar, ModJarMetadata mjm) {
+        var info = jar.moduleDataProvider().open(MODULE_INFO).orElse(null);
+        if (info != null) {
+            try {
+                var desc = ModuleDescriptor.read(info, jar::getPackages);
+                var all = new HashSet<>(jar.getPackages());
+                all.removeAll(desc.packages());
+                if (!all.isEmpty()) {
+                    var missing = all.stream().sorted().collect(Collectors.joining(", "));
+                    LOGGER.error("Invalid module-info, missing packages " + missing);
+                    throw new ModFileLoadingException("Invalid module-info, missing packages " + missing);
```

Comment:
No, this is for when the module-info.class is valid, but doesn't specify all the packages in the jar.
Leading to "why isn't my class loading?" Albeit, this would just be a invalid module that java would gladly load and cause ClassNotFoundException. I prefer this slightly more direct error. At the end of the day its a small (I'd guess <25 objects on average) HashSet difference, shouldn't be that big of a performance hit.